openedx/core/djangoapps/content_libraries/tests/test_api.py

```diff
@@ -13,8 +13,55 @@
     UsageKey,
 )
 from opaque_keys.edx.locator import LibraryLocatorV2
+from organizations.models import Organization
 
 from .. import api
+from ..constants import ALL_RIGHTS_RESERVED, COMPLEX
+from .base import requires_blockstore
+from openedx.core.lib.blockstore_api.methods import create_collection
+
+@requires_blockstore
+class ApiTest(TestCase):
+    """
+    Integration tests for the content libraries api functions
+    """
+
+    def test_bundle_links(self):
+        """
+        Test bundle links under various conditions
+        """
+        collection = create_collection('blockstore test collection 1')
+        org = Organization.objects.create(name='whatever', short_name='whatever')
+        content_library = api.create_library(
+            collection.uuid,
+            COMPLEX,
+            org,
+            'abcd1234',
+            'Test',
+            'example description',
+            True,
+            True,
+            ALL_RIGHTS_RESERVED,
+        )
+        assert api.get_bundle_links(content_library.key) == []
+
+        target_library = api.create_library(
+            collection.uuid,
+            COMPLEX,
+            org,
+            'abcd1235',
+            'Test2',
+            'example description 2',
+            True,
+            True,
+            ALL_RIGHTS_RESERVED,
+        )
+
+        target_key = target_library.key
+        # XXX this doesn't work
+        api.create_bundle_link(content_library.key, '1', target_key)
```

Comment:
This test runs fine for me?  Here's what I'm doing:

In the blockstore dir:
```
# Launch the edx.devstack.blockstore-test container
make testserver
```

In the `lms-shell`:
```
EDXAPP_RUN_BLOCKSTORE_TESTS=1 pytest openedx/core/djangoapps/content_libraries/tests/test_api.py::ApiTest
# ========================== 1 passed, 7 warnings in 4.30s ===========================
```

What problem are you seeing?

Reply:
I get this:

```
================================================================================== FAILURES ==================================================================================
_________________________________________________________________________ ApiTest.test_bundle_links __________________________________________________________________________

self = <openedx.core.djangoapps.content_libraries.tests.test_api.ApiTest testMethod=test_bundle_links>

    def test_bundle_links(self):
        """
        Test bundle links under various conditions
        """
        collection = create_collection('blockstore test collection 1')
        org = Organization.objects.create(name='whatever', short_name='whatever')
        content_library = api.create_library(
            collection.uuid,
            COMPLEX,
            org,
            'abcd1234',
            'Test',
            'example description',
            True,
            True,
            ALL_RIGHTS_RESERVED,
        )
        assert api.get_bundle_links(content_library.key) == []

        target_library = api.create_library(
            collection.uuid,
            COMPLEX,
            org,
            'abcd1235',
            'Test2',
            'example description 2',
            True,
            True,
            ALL_RIGHTS_RESERVED,
        )

        target_key = target_library.key
        # XXX this doesn't work
>       api.create_bundle_link(content_library.key, '1', target_key)

openedx/core/djangoapps/content_libraries/tests/test_api.py:63:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
openedx/core/djangoapps/content_libraries/api.py:1039: in create_bundle_link
    set_draft_link(draft.uuid, link_id, target_bundle_uuid, version)
openedx/core/lib/blockstore_api/methods.py:406: in set_draft_link
    api_request('patch', api_url('drafts', str(draft_uuid)), json={
openedx/core/lib/blockstore_api/methods.py:50: in api_request
    response.raise_for_status()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Response [400]>

    def raise_for_status(self):
        """Raises :class:`HTTPError`, if one occurred."""

        http_error_msg = ''
        if isinstance(self.reason, bytes):
            # We attempt to decode utf-8 first because some servers
            # choose to localize their reason strings. If the string
            # isn't utf-8, we fall back to iso-8859-1 for all other
            # encodings. (See PR #3538)
            try:
                reason = self.reason.decode('utf-8')
            except UnicodeDecodeError:
                reason = self.reason.decode('iso-8859-1')
        else:
            reason = self.reason

        if 400 <= self.status_code < 500:
            http_error_msg = u'%s Client Error: %s for url: %s' % (self.status_code, reason, self.url)

        elif 500 <= self.status_code < 600:
            http_error_msg = u'%s Server Error: %s for url: %s' % (self.status_code, reason, self.url)

        if http_error_msg:
>           raise HTTPError(http_error_msg, response=self)
E           requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://edx.devstack.blockstore-test:18251/api/v1/drafts/acae7ff6-d881-4428-a989-c81cd2a92edc

../venvs/edxapp/lib/python3.8/site-packages/requests/models.py:953: HTTPError
---------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------
['BundleVersion (b977f740-3d0a-49b1-a473-36d55eb61bf7, 0) referenced in Link 1 does not exist.']
```

Reply:
@swalladge I'm now almost as frustrated as you are :~

I started by adding this to the lx client code, so I could see what API calls are being invoked for your use case:

```diff
--- a/backend/labxchange/service_clients/edxapp/client.py
+++ b/backend/labxchange/service_clients/edxapp/client.py
@@ -269,6 +269,7 @@ class EdxAppClient:
         """
         Make an API call from Studio. Returns the parsed JSON response.
         """
+        log.info("CALL_STUDIO %s %s %s", method, path, kwargs)
         response = self.call_studio_raw(method, path, **kwargs)
         if response.status_code == 204:  # No content
             return None
```

I didn't make it all the way through, because my devstack won't let me save an Assignment for some reason, but maybe yours will?

1. Create library: POST /api/libraries/v2/ {'json': {'org': 'LabXchange', 'collection_uuid': '56789876-2111-4111-8111-1abe8c4ac9e0', 'slug': '9fc06738', 'title': 'LabXchange Content Library 9fc06738', 'description': '', 'allow_public_learning': True, 'allow_public_read': False}, 'edx_endpoint': '/api/libraries/v2/'}
2. Add `problem` block to library: POST /api/libraries/v2/lib:LabXchange:9fc06738/blocks/ {'json': {'block_type': 'problem', 'definition_id': '1'}, 'edx_endpoint': '/api/libraries/v2/{lib_key}/blocks/'}
3. Commit block: POST /api/libraries/v2/lib:LabXchange:9fc06738/commit/ {'edx_endpoint': '/api/libraries/v2/{lib_key}/commit/'}
4. (repeat for another problem block)
5. Create `lx_assignment` block: ... (here's where my devstack kacked it)


I recommend doing that first before trying to understand what I changed below -- because it's a bloody tangle.
By tracing through those calls, I started making these changes to your test, but it's not working, and I didn't have time tonight to figure out why.

```diff
diff --git a/openedx/core/djangoapps/content_libraries/tests/test_api.py b/openedx/core/djangoapps/content_libraries/tests/test_api.py
index a8f88a7245..d53dd876f9 100644
--- a/openedx/core/djangoapps/content_libraries/tests/test_api.py
+++ b/openedx/core/djangoapps/content_libraries/tests/test_api.py
@@ -61,30 +61,32 @@ class ApiTest(TestCase):
         )
         api.publish_changes(linked_library_1.key)

-        linked_library_2 = api.create_library(
-            collection.uuid,
-            COMPLEX,
-            org,
-            'abcd1236',
-            'Test3',
-            'example description 3',
-            True,
-            True,
-            ALL_RIGHTS_RESERVED,
+        assignment_block = api.create_library_block(
+            linked_library_1.key,
+            'lx_assignment',
+            'assignment1',
         )
-        api.publish_changes(linked_library_2.key)
+        api.publish_changes(linked_library_1.key)
+
+        problem_block_1 = api.create_library_block_child(
+            assignment_block.usage_key,
+            'problem',
+            'problem1',
+        )
+        api.publish_changes(linked_library_1.key)

-        # XXX: for me, creating this link fails at
-        # openedx/core/lib/blockstore_api/methods.py:406: in set_draft_link
-        # with error from blockstore (BundleVersion (****) referenced in Link one does not exist)
-        api.create_bundle_link(content_library.key, 'one', linked_library_1.key)
-        api.create_bundle_link(content_library.key, 'two', linked_library_2.key)
+        problem_block_2 = api.create_library_block_child(
+            assignment_block.usage_key,
+            'problem',
+            'problem2',
+        )
+        api.publish_changes(linked_library_1.key)

-        assert len(api.get_bundle_links(content_library.key)) == 2
+        assert len(api.get_bundle_links(assignment_block.usage_key)) == 2

-        api.delete_library(linked_library_1.key)
+        api.delete_library_block(problem_block_1.library_key, remove_from_parent=False)

-        assert len(api.get_bundle_links(content_library.key)) == 1
+        assert len(api.get_bundle_links(assignment_block.usage_key)) == 1

         # TODO: assert bundle links content
```

From that, I get this stack trace:

```
___________________________________________________________________________________ ApiTest.test_bundle_links ____________________________________________________________________________________
self = <openedx.core.djangoapps.content_libraries.tests.test_api.ApiTest testMethod=test_bundle_links>

    def test_bundle_links(self):
        """
        Test bundle links under various conditions
        """
        collection = create_collection('blockstore test collection 1')
        org = Organization.objects.create(name='whatever', short_name='whatever')
        content_library = api.create_library(
            collection.uuid,
            COMPLEX,
            org,
            'abcd1234',
            'Test',
            'example description',
            True,
            True,
            ALL_RIGHTS_RESERVED,
        )
        api.publish_changes(content_library.key)

        assert api.get_bundle_links(content_library.key) == []

        linked_library_1 = api.create_library(
            collection.uuid,
            COMPLEX,
            org,
            'abcd1235',
            'Test2',
            'example description 2',
            True,
            True,
            ALL_RIGHTS_RESERVED,
        )
        api.publish_changes(linked_library_1.key)

        assignment_block = api.create_library_block(
            linked_library_1.key,
            'lx_assignment',
            'assignment1',
        )
        api.publish_changes(linked_library_1.key)

>       problem_block_1 = api.create_library_block_child(
            assignment_block.usage_key,
            'problem',
            'problem1',
        )

openedx/core/djangoapps/content_libraries/tests/test_api.py:71:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
openedx/core/djangoapps/content_libraries/api.py:875: in create_library_block_child
    parent_block.save()
../venvs/edxapp/lib/python3.8/site-packages/xblock/mixins.py:251: in save
    self.runtime.save_block(self)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <openedx.core.djangoapps.xblock.runtime.blockstore_runtime.BlockstoreXBlockRuntime object at 0x7f1b66537430>
block = <AssignmentBlockWithMixins @4274 parent=???, name=None, tags=[], children=[LibraryUsageLocatorV2(LibraryLocatorV2('whatever', 'abcd1235'), 'problem', 'problem1')], display_name='Assignment'>

    def save_block(self, block):
        """
        Save any pending field data values to Blockstore.

        This gets called by block.save() - do not call this directly.
        """
        if not self.system.authored_data_store.has_changes(block):
            return  # No changes, so no action needed.
        definition_key = block.scope_ids.def_id
        if definition_key.draft_name is None:
>           raise RuntimeError(
                "The Blockstore runtime does not support saving changes to blockstore without a draft. "
                "Are you making changes to UserScope.NONE fields from the LMS rather than Studio?"
            )
E           RuntimeError: The Blockstore runtime does not support saving changes to blockstore without a draft. Are you making changes to UserScope.NONE fields from the LMS rather than Studio?

openedx/core/djangoapps/xblock/runtime/blockstore_runtime.py:124: RuntimeError
```

I _think_ this means that instead of using the `assignment_block.usage_key` in the `api.create_library_block_child` call, you need a draft version of the assignment instead? But I really don't know.

Hoping this gives you somewhere to start!

Reply:
I think it's time to ask if we really need to add a test for this.... :P

Reply:
@swalladge  It's an intricate and problematic bug, so yes, we need tests for it.

Have you had any luck following the API calls that LX uses? Since you can reliably reproduce the issue there, you should be able to do the same here using the api calls directly, right?

Reply:
Sorted it out at last!! :D

Reply:
![1630559263](https://user-images.githubusercontent.com/9714796/131785279-3bfc006f-de04-46e9-b0d6-0229c16744ae.png)

Here's the coverage report of the function to prove it. :D

Reply:
You are a rock start @swalladge ! Thank you so much for persisting here!